app/backend/src/couchers/servicers/auth.py

```diff
@@ -170,102 +153,194 @@ def _username_available(self, username):
-
-    def SignupTokenInfo(self, request, context):
-        """
-        Returns the email for a given SignupToken (which will be shown on the UI on the singup form).
-        """
-        logger.debug(f"Signup token info for {request.signup_token=}")
+    def SignupFlow(self, request, context):
         with session_scope() as session:
-            signup_token = (
-                session.query(SignupToken)
-                .filter(SignupToken.token == request.signup_token)
-                .filter(SignupToken.is_valid)
-                .one_or_none()
-            )
-            if not signup_token:
-                context.abort(grpc.StatusCode.NOT_FOUND, errors.INVALID_TOKEN)
+            if request.email_token:
+                # the email token can either be for verification or just to find an existing signup
+                flow = (
+                    session.query(SignupFlow)
+                    .filter(SignupFlow.email_verified == False)
+                    .filter(SignupFlow.email_token == request.email_token)
+                    .filter(SignupFlow.token_is_valid)
+                    .one_or_none()
+                )
+                if flow:
+                    # find flow by email verification token and mark it as verified
+                    flow.email_verified = True
+                    flow.email_token = None
+                    flow.email_token_created = None
+                    flow.email_token_expiry = None
+
+                    session.flush()
+                else:
+                    # just try to find the flow by flow token, no verification is done
+                    flow = session.query(SignupFlow).filter(SignupFlow.flow_token == request.email_token).one_or_none()
+                    if not flow:
+                        context.abort(grpc.StatusCode.NOT_FOUND, errors.INVALID_TOKEN)
             else:
-                return auth_pb2.SignupTokenInfoRes(email=signup_token.email)
+                if not request.flow_token:
+                    # fresh signup
+                    if not request.HasField("basic"):
+                        context.abort(grpc.StatusCode.INVALID_ARGUMENT, errors.SIGNUP_FLOW_BASIC_NEEDED)
+                    # TODO: unique across both tables
+                    existing_user = session.query(User).filter(User.email == request.basic.email).one_or_none()
+                    if existing_user:
+                        context.abort(grpc.StatusCode.FAILED_PRECONDITION, errors.SIGNUP_FLOW_EMAIL_TAKEN)
+                    existing_flow = (
+                        session.query(SignupFlow).filter(SignupFlow.email == request.basic.email).one_or_none()
+                    )
+                    if existing_flow:
+                        send_signup_email(existing_flow)
+                        session.commit()
+                        context.abort(grpc.StatusCode.FAILED_PRECONDITION, errors.SIGNUP_FLOW_EMAIL_STARTED_SIGNUP)
```

Comment:
I think we should allow several people starting to sign up using the same email address, and the email that gets clicked will "win" and the other partially signed up will be discarded.

Reply:
If we allow duplicate usernames, your username might be stolen in front of your eyes although it was was ok in the beginning of the signup flow. How should we handle usernames?

Reply:
I think it'd be good to check in the backend to make sure the usernames are approximately unique across both tables (though there will still be a slight race condition), just to prevent this as a courtesy to the user.

But otherwise I think it's okay to just have it unique in both tables and the database can enforce that.